lib/spack/spack/fetch_strategy.py

```diff
@@ -421,6 +430,79 @@ def fetch(self):
         tty.msg('Using cached archive: %s' % path)
 
 
+class PyPiFetchStrategy(URLFetchStrategy):
+    """Fetches packages directly from PyPi using XML-RPC to query the API."""
+
+    enabled = True
+    required_attributes = ['pypi']
+
+    def __init__(self, pypi=None, digest=None, **kwargs):
+        # If pypi or digest are provided in the kwargs, then prefer
+        # those values.
+        self.pypi = kwargs.get('pypi', None)
+        if not self.pypi:
+            self.pypi = pypi
```

Comment:
I know something is fishy here, but that's how `URLFetchStrategy` did things, so I copied it just to be safe. Honestly, all of the FetchStrategy logic needs to be reworked. As you've noticed, it's very opaque and not well documented.